### ISSUE TITLE
No detection of the end of my track

### ISSUE THREAD

**Issue by merlinletekos** [sedmelluq/lavaplayer#113]
So when a start to play my first track, all work fine, but when the tracks were finished, nothing, the other track doesn't load. This program worked fine during... the 14/11/2017 so now, I try to change something, but, nothing, you can help me? and if other people have a the same problem you can search with me a solution? 
```
public void nextTrack(){
    if(tracks.isEmpty()){
        if(player.getGuild().getAudioManager().getConnectedChannel() != null) {
            player.getAudioPlayer().destroy();
            player.getGuild().getAudioManager().closeAudioConnection();
        }
        return;
    }
    System.out.println(AudioTrackEndReason.class.getName());
   player.getAudioPlayer().startTrack(tracks.poll(), false);
}

public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
    if (endReason.mayStartNext) {
        nextTrack();
        System.out.println("========");
    }
}
```

[The repo of my bot](https://github.com/LordAll/LordBot)

**Comment by shikhir-arora** [comment]
We need more logs and info than this to help you. It's a semi-common issue; as in handling `TrackEnd` events properly on Lavalink is mostly up to the client. But, even then you have to properly catch all the edge cases so, it could be that. 

I know you said you use d.js-lavalink so I'm familiar with it and I've never had that problem (only once or twice when it was isolated and I had known why).

So you're using Lavalink, can you post your debug log on the server end? Likely there's not much there, but it's good to have.

So if you have (and you should if you don't) a way to evaluate directly with your bot, then you can evaluate something like this when the track ends -

`client.player.get(message.guild.id)` - this will give you an object with a lot of properties we can look at. If this errors as undefined then you can go from there. If it doesn't, then post back and you can easily trigger the event yourself and expand it. If this is happening more than like a few isolated times and you haven't changed the library it's either skwmrbing in your code or maybe a rare edge case I never experienced.

**Comment by Bacon1153** [comment]
Okay so I am experiencing the bug however it's a lot worse now. The end event is not firing at all, and it's now a lot more frequent. I can call the end event myself though by going `bot.player.get("356197359153315840")._events.end("REPLACED");` and it works as normal. The player returns:
```
Player {
  domain: null,
  _events: [Object],
  _eventsCount: 2,
  _maxListeners: undefined,
  options: [Object],
  id: '356197359153315840',
  client: [Object],
  manager: [Object],
  node: [Object],
  channel: '356197359715221514',
  playing: true,
  state: {},
  track: 'QAAAigIAK1NuYWlsJ3MgSG91c2UgLSBQaXhlbCBEcmVhbSBbVGFzdHkgUmVsZWFzZV0ABVRhc3R5AAAAAAADZxgAC0tZblM0ZG93bWJ3AAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9S1luUzRkb3dtYncAB3lvdXR1YmUAAAAAAAAAAA==',
  timestamp: 1527993825428,
  queue: [Array],
  timer: null }
```
Although it has a depth of 0, so if you need anything more specific just let me know.

Here's the debug.log: https://hastebin.com/raw/pemoloqoci

And my code for the play command: https://hastebin.com/umecarotuh.js
And it may be linked to how terrible my code is. I really need to get around to rewriting it, but it doesn't explain as to why it used to work 100% of the time, and now only half... so. Also, thanks for taking interest on this issue. I hope we can get it resolved.

**Comment by Bacon1153** [comment]
I am on my phone too atm. I'm pretty sure we can dismiss my issue now as I'm sure it's an issue with my code. It's pretty interesting though that my code has been working flawlessly from February to half way through April. I'll make a rewrite and get back to you later. And also, this is quite hard for me to understand from you reply -- You want me to use stop instead of leave?

**Comment by shikhir-arora** [comment]
I haven't read your code or ran it, I'm not saying that will fix it because it very likely won't, but, leave calls destroy and stop doesn't, as in when a track is stopped (or TrackEnd event for example) you can go to the next track without destroying it. It's all in the lib.

https://imgur.com/a/4V9Fw6j

**Comment by Bacon1153** [comment]
Okay so I just got around to doing a "rewrite" and by no means is it great, but it has some major improvements. I have not been able to reproduce the end event bug yet with it. https://hastebin.com/uxadafetoq.js

One thing I noticed though is that the error event is not working at all. It's not even working with my old code where I know FOR SURE it was.

Also, I think you misread my code. I don't destroy the player after playing a song, only when there's no more songs in queue.
